rllib/examples/metrics/custom_metrics_in_env_runners.py

```diff
@@ -0,0 +1,339 @@
+"""Example of adding custom metrics to the results returned by `EnvRunner.sample()`.
+
+We use the `MetricsLogger` class, which RLlib provides inside all its components (only
+when using the new API stack through
+`config.api_stack(_enable_rl_module_and_learner=True,
+_enable_env_runner_and_connector_v2=True)`),
+and which offers a unified API to log individual values per iteration, per episode
+timestep, per episode (as a whole), per loss call, etc..
+`MetricsLogger` objects are available in all custom API code, for example inside your
+custom `Algorithm.training_step()` methods, custom loss functions, custom callbacks,
+and custom EnvRunners.
+
+This example:
+    - demonstrates how to write a custom Callbacks subclass, which overrides some
+    EnvRunner-bound methods, such as `on_episode_start`, `on_episode_step`, and
+    `on_episode_env`.
+    - shows how to temporarily store per-timestep data inside the currently running
+    episode within the EnvRunner (and the callback methods).
+    - shows how to extract this temporary data again when the episode is done to
+    further process it into a reportable metric.
+    - explains how to use the `MetricsLogger` API to create and log different metrics
+    to the final Algorithm's iteration output. These include - but are not limited to -
+    a 2D heatmap (image) per episode, an average per-episode metric (over a sliding
+    window of 200 episodes), a maximum per-episode metric (over a sliding window of 100
+    episodes), and an EMA-smoothed metric.
+
+In this script, we define a custom `DefaultCallbacks` class and then override some of
+its methods in order to define custom behavior during episode sampling. In particular in
+this example, we add custom metrics to the Algorithm's published result dict (once per
+iteration) before it is sent back to Ray Tune (and possibly a WandB logger).
+
+For demonstration purposes only, we log the following custom metrics:
+- A 2D heatmap showing the frequency of certain y/x-locations of Ms Pacman during an
+episode. We create and log a separate heatmap per episode and limit the number of
+heatmaps reported back to the algorithm to 10.
+- The maximum per-episode distance travelled by Ms Pacman over a sliding window of 100
+episodes.
+- The average per-episode distance travelled by Ms Pacman over a sliding window of 200
+episodes.
+- The EMA-smoothed number of lives of Ms Pacman at each timestep (across all episodes).
+
+
+How to run this script
+----------------------
+`python [script file name].py --enable-new-api-stack --wandb-key [your WandB key]
+--wandb-projecy [some project name]`
```

Comment:
"--wandb-projecy" -> "--wandb-project"

Reply:
The same options occur below in "For logging ..."

Reply:
Yeah, that's ok. The statement below is the generic enable-logging statement, the one above are the recommended args for this particular script.